backend/pps/sch.py

```diff
@@ -72,3 +79,38 @@ def mp_consumed_per_cycle(self, character_stats, caster_tax, succ, adlo, ed, rez
         # Energy Drain
         result -= 500 * ed * cycle / 60
         return result
+    
+    def get_total_potency_flexible_time(self, character, num_seconds, caster_tax):        
```

Comment:
i really really think that if you're doing the full rewrite you should break things up into separate functions for determining what spells are used, then a function that adds up the potency.

reason being that for custom rotations, clearly we can just call the second function and pass in whatever that rotation is. 

tbh i wonder if using an iterator to generate the "default" rotation wouldn't be horrible:
hardcode the gcd 4 chain opener as a starting point. then:

if the dot has < 1 gcd remaining on its timer (or is not on the boss) and aetherflow cd is not between 1 and 2 gcds out, your next gcd should be bio. if aetherflow is off cd then the ogcds are af, ed. otherwise, if swift is available, your ogcds are ed, swift. if swift isn't available then it's just ed.

if the previous gcd was bio with no aetherflow, then if swift was used, your gcd is broil followed by ed. otherwise it's ruin 2 followed by ed. 

i think the only questionable thing is future uses of dissipation because you need to burn aetherflow stacks before dissipation comes up... mark that one as a todo? it's only 30 potency gain every 3 minutes lmao.